### ISSUE TITLE
Promise.all return type incorrect when some promise types are unions of Promise types, rather than Promise of union of types.

### ISSUE THREAD

**Issue by UselessPickles** [microsoft/TypeScript#37216]
**TypeScript Version:**  3.8.3 (regression - does not occur in 3.7)

**Search Terms:** Promise.all return type

**Code**

```ts
declare const conditional: boolean;

async function foo(): Promise<void> {
    // Note: promiseA cannot possibly resolve to undefined
    const promiseA = Promise.resolve("hello");
    const promiseB = conditional ? Promise.resolve(10) : Promise.resolve(undefined);

    const [resultA, resultB] = await Promise.all([promiseA, promiseB]);

    // Unexpected error here: resultA is possibly undefined
    alert(resultA.toUpperCase());
}```

**Expected behavior:**
Type of `resultA` should be `string`;

**Actual behavior:**
Type of `resultA` is `string | undefined`;

**Workaround:**
Tack on a `as const` to the tuple of promises being passed to `Promise.all()`:
```ts
    const [resultA, resultB] = await Promise.all([promiseA, promiseB] as const);
```

**Playground Link:** http://www.typescriptlang.org/play/index.html#code/CYUwxgNghgTiAEYD2A7AzgF0a4BLDuqUEAXPAEZJIQhQoDcAUI1GgJ4pjwBmArpwVQ8qACgCUZAAowkAW1xoQAHgBuSXMAB88AN6N4B+AHoj8AHJIMIMgAcZ8xQEFEdFJfg2kaNLnIQ28HBo1CoIGEjw-KDcuCggwPqGyOhYdnIKIM4AvPDS6YoAdEEhICIARAAWIBAQSGViTIbYKR72GQBC8DnJeIIoxPAA-LlthcUQoSIAjAAMYvBSoyBFIMETpVEgMXHADcxNyZjwANpBvBAYjgA0gavnGO0Aul3wUADuUPgj+cvEECLHNIOTI3IEdR57RIGEzwACqcQAHjZwFZgPAQDAZDB4FU4GQzhdnAoPF4fH4ApttvEoa8aDAMCICZcCuFYTZkTAAMKsUpiPYAX2YQA

**Comment by UselessPickles** [comment]
I think I just realized that `undefined` has nothing to do with this, but instead that one of the promises' type is a union of different Promise types?

**Comment by nmain** [comment]
> **Search Terms:** Promise.all return type

Probable duplicate of https://github.com/microsoft/TypeScript/issues/35626
Probable duplicate of https://github.com/microsoft/TypeScript/issues/34937
Probable duplicate of https://github.com/microsoft/TypeScript/issues/33562

I think these all trace back to https://github.com/microsoft/TypeScript/pull/33057, which is [my fault](https://github.com/microsoft/TypeScript/issues/31179#issuecomment-488275275), so I am sorry.

**Comment by UselessPickles** [comment]
@nmain Thanks. I don't know what's wrong with my eyes. Looking at the search results again it seems obvious that those were likely duplicates I should have looked at.
